programs/ssl/ssl_test_lib.c

```diff
@@ -0,0 +1,283 @@
+    {
+        first_try = 0;
+        return( MBEDTLS_ERR_SSL_WANT_WRITE );
+    }
+
+    ret = mbedtls_net_send( ctx, buf, len );
+    if( ret != MBEDTLS_ERR_SSL_WANT_WRITE )
+        first_try = 1; /* Next call will be a new operation */
+    return( ret );
+}
+
+#define HEX2NUM( c )                    \
+        if( c >= '0' && c <= '9' )      \
+            c -= '0';                   \
+        else if( c >= 'a' && c <= 'f' ) \
+            c -= 'a' - 10;              \
+        else if( c >= 'A' && c <= 'F' ) \
+            c -= 'A' - 10;              \
+        else                            \
+            return( -1 );
+
+/*
+ * Convert a hex string to bytes.
+ * Return 0 on success, -1 on error.
+ */
+int mbedtls_ssl_test_unhexify( unsigned char *output, const char *input, size_t *olen )
+{
+    unsigned char c;
+    size_t j;
+
+    *olen = strlen( input );
+    if( *olen % 2 != 0 || *olen / 2 > MBEDTLS_PSK_MAX_LEN )
+        return( -1 );
+    *olen /= 2;
+
+    for( j = 0; j < *olen * 2; j += 2 )
+    {
+        c = input[j];
+        HEX2NUM( c );
+        output[ j / 2 ] = c << 4;
+
+        c = input[j + 1];
+        HEX2NUM( c );
+        output[ j / 2 ] |= c;
+    }
+
+    return( 0 );
+}
+
+int mbedtls_ssl_test_force_ciphersuite( int force_ciphersuite,
+                                        int transport,
+                                        int min_version,
+                                        int max_version,
+                                        int *arc4 )
+{
+    const mbedtls_ssl_ciphersuite_t *ciphersuite_info;
+    ciphersuite_info = mbedtls_ssl_ciphersuite_from_id( force_ciphersuite );
+
+    if( max_version != -1 &&
+        ciphersuite_info->min_minor_ver > max_version )
+    {
+        mbedtls_printf("forced ciphersuite not allowed with this protocol version\n");
+        return( 2 );
+    }
+    if( min_version != -1 &&
+        ciphersuite_info->max_minor_ver < min_version )
+    {
+        mbedtls_printf("forced ciphersuite not allowed with this protocol version\n");
+        return( 2 );
+    }
+
+    /* If the server selects a version that's not supported by
+     * this suite, then there will be no common ciphersuite... */
+    if( max_version == -1 ||
+        max_version > ciphersuite_info->max_minor_ver )
+    {
+        max_version = ciphersuite_info->max_minor_ver;
+    }
+    if( min_version < ciphersuite_info->min_minor_ver )
+    {
+        min_version = ciphersuite_info->min_minor_ver;
+        /* DTLS starts with TLS 1.1 */
+        if( transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM &&
+            min_version < MBEDTLS_SSL_MINOR_VERSION_2 )
+            min_version = MBEDTLS_SSL_MINOR_VERSION_2;
+    }
+
+    /* Enable RC4 if needed and not explicitly disabled */
+    if( ciphersuite_info->cipher == MBEDTLS_CIPHER_ARC4_128 )
+    {
+        if( *arc4 == MBEDTLS_SSL_ARC4_DISABLED )
+        {
+            mbedtls_printf("forced RC4 ciphersuite with RC4 disabled\n");
+        return( 2 );
```

Comment:
Indentation issue.

Reply:
Fixed. Since there were no other review comments, I amended the offending commit. The old version is at https://github.com/gilles-peskine-arm/mbedtls/tree/ssl_test_lib-development-2.11-1